### DISCUSSION TITLE
Add ODBC support

### DISCUSSION THREAD

**Discussion by Angemell** [godotengine/godot-proposals#4609] [General Discussions]
Hello Everyone: 

I'm doing the labor of recreating this (https://github.com/godotengine/godot/issues/5213), somehow forgotten feature, that is desired by some of us, developers in Godot. I'm, transcribing the original post by [silverkorn](https://github.com/silverkorn) as I believe it is as clear and straight to the point as is needed:

> Greetings,
> In many context, specially those with a lot of information to store such as user information, progress, logins & scoreboards, databases are fundamental to those applications/games.
> Unfortunately, there's no support for larger scaled databases in Godot and the current format is extremely limited.
> This is why I suggest the implementation of ODBC with, as suggestion, [nanodbc](http://lexicalunit.github.io/nanodbc/).
> It is MIT licensed, light and cross-platform.
> Why do I suggested ODBC over any other specific method/protocol?
> Simply because it's supported by most of database softwares (e.g.: Microsoft SQL, MySQL, PostgreSQL, SQLite, etc.) with a generic  syntax and it's platform independent.
> Let me know your thoughts about it.
> Thanks! 😄

p.d. please let me know if I'm posting this request in the proper manner. Thanks :)

**Comment by Calinou** [comment]
I think this is still better implemented as an add-on. Godot 4.0 will have [GDExtension](https://godotengine.org/article/introducing-gd-extensions) which makes developing this kind of add-on much easier.

**Reply by Angemell** [reply]
ok, ok .. sounds great; should I keep this post here? or should I re-post it somewhere else so it makes some echo, and might get attended by some fellow Godot's contributors?

**Reply by Calinou** [reply]
Discussions are made to be open-ended, so they are never closed (unless they're locked).

You can wait and see if anyone wants to tackle this, but don't expect anyone to do work for you for free. Typically, people will work on something if they need it in their own project 🙂 
I don't plan to work on this personally.

**Reply by Angemell** [reply]
I see ... well, perhaps having this open discussion might inspire someone more skilled than me (that's for sure), so that development will be of use for all of us in such a need. Btw... is it of "good taste" to negotiate here directly on GitHub if someone has the skills, and is willing to do the job? (of course, if it is reasonable for my limited budget).

Thanks a lot for the kind and detailed reply :)

**Reply by Calinou** [reply]
> Btw... is it of "good taste" to negotiate here directly on GitHub if someone has the skills, and is willing to do the job? (of course, if it is reasonable for my limited budget).

Sure, just be aware that we can't make any guarantees as to the reliability of said work.

**Reply by Angemell** [reply]
ok, ok... I'll be aware if someone seeing this thread raises the hand in dealing with the job. I guess that the risk of a scam (which is different of the "no warranties policy”), is at hand, but we have to appeal for the good ethos of our developer fellows.

Thanks a lot for the kind follow of my concerns.. it makes a great difference! (other communities just tend to ignore you).

**Comment by Angemell** [comment]
Hello fellow Godot Developers:

I've come to the following (at least provisional), alternative for using ODBC at a Godot Project:

* Install Godot for Mono (64 bits preferably, but it will work with 32 bits as well).
* Configure your default Script Editor to **Visual Studio Code** (it all seems that Visual Studio 2019/2022 is still buggy to be reliable): 
Editor -> Editor Settings -> General Tab -> Mono -> Editor -> External Editor: Visual Studio Code

![Editor](https://user-images.githubusercontent.com/99204519/173222404-066022dc-05d3-4248-9cec-a46b4e048f5a.png)

* At Visual Studio Code's Terminal: Install [System.Data.Odbc](https://www.nuget.org/packages/System.Data.Odbc/) via .NET CLI: > dotnet add package System.Data.Odbc --version 6.0.0
![VSC-Terminal](https://user-images.githubusercontent.com/99204519/173223265-d2e7cf98-2b9e-4387-bf88-dd82327fafb2.png)

* Be sure that your csproj.file contains the following lines:
```
  <ItemGroup>
        <PackageReference Include="System.Data.Odbc" Version="6.0.0"/>
  </ItemGroup>
```
  
![csproj](https://user-images.githubusercontent.com/99204519/173222525-eb826737-8fb7-4c48-bbd6-08ea205e6e53.png)

* Include the necessary _using’s_ and basic ODBC code; here it goes my snippet:

```
using Godot;
using System;
using System.Configuration;
using System.Data;
using System.Data.Odbc;

public class MainNode : Node2D {
    DataTable dtAux = new DataTable();
    OdbcDataReader dbReader;

    // Called when the node enters the scene tree for the first time.
    public override void _Ready() {
        OdbcConnection conn = new OdbcConnection("Driver={Microsoft Access Driver (*.mdb, *.accdb)};DBQ=C:\\RPG\\Rpg.accdb;Uid=Admin;Pwd=;");  //Access DB (Office @ 64 bits)
        //OdbcConnection conn = new OdbcConnection("Driver={PostgreSQL ANSI};Server=myserver.net;Port=5432;Database=RPGDemo;Uid=postgres;Pwd=mypassword;"); //PostgreSQL Remote Server 
        OdbcCommand comm = new OdbcCommand("SELECT * FROM items", conn);
        try { conn.Open();
              dbReader = comm.ExecuteReader();
              dtAux.Load(dbReader);

              GD.Print("Conexion Exitosa!");
              GD.Print(dtAux.Rows[0]["item_name"].ToString());
              GD.Print(dtAux.Rows[1]["item_name"].ToString());
              conn.Close();
        }
        catch (Exception ex) { GD.Print(ex); }
    }
}
```
![Access](https://user-images.githubusercontent.com/99204519/173222792-406ea5ff-759a-449e-9256-c628778e66a3.png)

![PostgreSQL](https://user-images.githubusercontent.com/99204519/173222797-6c717462-5852-4447-9e15-46a9a568ab63.png)

  * Run your project at Godot:

![GodotSuccess1](https://user-images.githubusercontent.com/99204519/173223371-40847eca-ab1c-472d-beea-437e068afc9d.png)
   
  I hope this is of help for anyone coming for a provisional alternative in case someone can give us a more professional solution.
  
  Best wishes to everyone :)
  Regards,

p.s. the PostgreSQL ODBC Driver can be downloaded [here](https://www.postgresql.org/download/) .
Direct link of [PostgreSQL ODBC Driver for Windows](https://www.postgresql.org/ftp/odbc/versions/msi/)  (scroll down and look for the lastest version; in my case v13.0.2)
